http3/conn.go

```diff
@@ -0,0 +1,530 @@
+package http3
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/lucas-clemente/quic-go"
+	"github.com/lucas-clemente/quic-go/quicvarint"
+)
+
+const (
+	maxBufferedStreams   = 10
+	maxBufferedDatagrams = 10
+)
+
+// Conn is a base HTTP/3 connection.
+// Callers should use either ServerConn or ClientConn.
+type Conn interface {
+	// Settings returns the HTTP/3 settings for this side of the connection.
+	Settings() Settings
+
+	// PeerSettings returns the peer’s HTTP/3 settings.
+	// This will block until the peer’s settings have been received.
+	PeerSettings() (Settings, error)
+}
+
+// ServerConn is a server connection. It accepts and processes HTTP/3 request streams.
+type ServerConn interface {
+	Conn
+	AcceptRequestStream(context.Context) (RequestStream, error)
+}
+
+// ClientConn is a client connection. It opens and processes HTTP/3 request streams.
+type ClientConn interface {
+	Conn
+	OpenRequestStream(context.Context) (RequestStream, error)
+}
+
+// webTransportConn is an internal interface for implementing WebTransport.
+type webTransportConn interface {
+	acceptStream(context.Context, SessionID) (quic.Stream, error)
+	acceptUniStream(context.Context, SessionID) (quic.ReceiveStream, error)
+	openStream(SessionID) (quic.Stream, error)
+	openStreamSync(context.Context, SessionID) (quic.Stream, error)
+	openUniStream(SessionID) (quic.SendStream, error)
+	openUniStreamSync(context.Context, SessionID) (quic.SendStream, error)
+	readDatagram(context.Context, SessionID) ([]byte, error)
+	writeDatagram(SessionID, []byte) error
+}
+
+type connection struct {
```

Comment:
To elaborate, you can have a `bool` here which tells you if this is a client or a server connection. Then you don't need to ask the `Session` for its `Perspective()`. You just have to set that `bool` once when you initialize this struct.

Reply:
How would Open or Accept know to error on the wrong perspective if it's not exposed at the quic layer?